README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# `codexchain`

<!-- badges: start -->

[![R-CMD-check](https://github.com/andrewallenbruce/codexchain/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/andrewallenbruce/codexchain/actions/workflows/R-CMD-check.yaml)
[![code
size](https://img.shields.io/github/languages/code-size/andrewallenbruce/codexchain.svg)](https://github.com/andrewallenbruce/codexchain)
[![Codecov test
coverage](https://codecov.io/gh/andrewallenbruce/codexchain/branch/master/graph/badge.svg)](https://app.codecov.io/gh/andrewallenbruce/codexchain?branch=master)
[![last
commit](https://img.shields.io/github/last-commit/andrewallenbruce/codexchain.svg)](https://github.com/andrewallenbruce/codexchain/commits/master)
[![Project Status: WIP - Initial development is in progress, but there
has not yet been a stable, usable release suitable for the
public.](https://www.repostatus.org/badges/latest/wip.svg)](https://www.repostatus.org/#wip)
[![License: GPL (\>=
3)](https://img.shields.io/badge/license-GPL%20(%3E=%203)-blue.svg)](https://cran.r-project.org/web/licenses/GPL%20(%3E=%203))
<!-- badges: end -->

## Installation

You can install the development version of codexchain from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("andrewallenbruce/codexchain")

# install.packages("remotes")
remotes::install_github("andrewallenbruce/codexchain")
```

``` r
# load library
library(codexchain)
```

## National Library of Medicine’s **ICD-10-CM 2023** API

The *International Classification of Diseases, 10th Revision, Clinical
Modification* is the American modification of the World Health
Organization’s ICD-10, a medical coding system for classifying diagnoses
and reasons for healthcare visits.

Return the seven ICD-10-CM codes beginning with “A15”

``` r
codex_icd10(code = "A15") |> 
     gluedown::md_table()
```

| icd_10_cm_code | icd_10_cm_term                               |
|:---------------|:---------------------------------------------|
| A15.0          | Tuberculosis of lung                         |
| A15.4          | Tuberculosis of intrathoracic lymph nodes    |
| A15.5          | Tuberculosis of larynx, trachea and bronchus |
| A15.6          | Tuberculous pleurisy                         |
| A15.7          | Primary respiratory tuberculosis             |
| A15.8          | Other respiratory tuberculosis               |
| A15.9          | Respiratory tuberculosis unspecified         |

<br>

Return the first 20 ICD-10-CM codes associated with tuberculosis

``` r
codex_icd10(term = "tuber", 
            limit = 20) |> 
            gluedown::md_table()
```

| icd_10_cm_code | icd_10_cm_term                                    |
|:---------------|:--------------------------------------------------|
| A15.0          | Tuberculosis of lung                              |
| A15.4          | Tuberculosis of intrathoracic lymph nodes         |
| A15.5          | Tuberculosis of larynx, trachea and bronchus      |
| A15.6          | Tuberculous pleurisy                              |
| A15.7          | Primary respiratory tuberculosis                  |
| A15.8          | Other respiratory tuberculosis                    |
| A15.9          | Respiratory tuberculosis unspecified              |
| A17.0          | Tuberculous meningitis                            |
| A17.1          | Meningeal tuberculoma                             |
| A17.81         | Tuberculoma of brain and spinal cord              |
| A17.82         | Tuberculous meningoencephalitis                   |
| A17.83         | Tuberculous neuritis                              |
| A17.89         | Other tuberculosis of nervous system              |
| A17.9          | Tuberculosis of nervous system, unspecified       |
| A18.01         | Tuberculosis of spine                             |
| A18.02         | Tuberculous arthritis of other joints             |
| A18.03         | Tuberculosis of other bones                       |
| A18.09         | Other musculoskeletal tuberculosis                |
| A18.10         | Tuberculosis of genitourinary system, unspecified |
| A18.11         | Tuberculosis of kidney and ureter                 |

<br>

Return the two ICD-10-CM codes associated with pleurisy

``` r
codex_icd10(term = "pleurisy") |> 
           gluedown::md_table()
```

| icd_10_cm_code | icd_10_cm_term       |
|:---------------|:---------------------|
| R09.1          | Pleurisy             |
| A15.6          | Tuberculous pleurisy |

<br>

When searching for ICD-10-CM codes by letter, you must set the `field`
param to “code” or it will search for *terms* containing the letter as
well. For example, without `field = "code"`, the following returns terms
containing the letter “Z”:

``` r
codex_icd10(code = "Z", 
            limit = 10) |> 
   gluedown::md_table()
```

| icd_10_cm_code | icd_10_cm_term                                                        |
|:---------------|:----------------------------------------------------------------------|
| A28.8          | Other specified zoonotic bacterial diseases, not elsewhere classified |
| A28.9          | Zoonotic bacterial disease, unspecified                               |
| A92.5          | Zika virus disease                                                    |
| B02.0          | Zoster encephalitis                                                   |
| B02.1          | Zoster meningitis                                                     |
| B02.30         | Zoster ocular disease, unspecified                                    |
| B02.31         | Zoster conjunctivitis                                                 |
| B02.32         | Zoster iridocyclitis                                                  |
| B02.33         | Zoster keratitis                                                      |
| B02.34         | Zoster scleritis                                                      |

<br>

With `field = "code"`, it returns only codes containing “Z”:

``` r
codex_icd10(code = "z", 
            field = "code", 
            limit = 10) |> 
      gluedown::md_table()
```

| icd_10_cm_code | icd_10_cm_term                                                                           |
|:---------------|:-----------------------------------------------------------------------------------------|
| Z00.00         | Encounter for general adult medical examination without abnormal findings                |
| Z00.01         | Encounter for general adult medical examination with abnormal findings                   |
| Z00.110        | Health examination for newborn under 8 days old                                          |
| Z00.111        | Health examination for newborn 8 to 28 days old                                          |
| Z00.121        | Encounter for routine child health examination with abnormal findings                    |
| Z00.129        | Encounter for routine child health examination without abnormal findings                 |
| Z00.2          | Encounter for examination for period of rapid growth in childhood                        |
| Z00.3          | Encounter for examination for adolescent development state                               |
| Z00.5          | Encounter for examination of potential donor of organ and tissue                         |
| Z00.6          | Encounter for examination for normal comparison and control in clinical research program |

<br> <br>

## BMI Coding Example

``` r
codex_icd10(term = "bmi adult", 
            limit = 30) |> 
  gluedown::md_table()
```

| icd_10_cm_code | icd_10_cm_term                               |
|:---------------|:---------------------------------------------|
| Z68.45         | Body mass index \[BMI\] 70 or greater, adult |
| Z68.1          | Body mass index \[BMI\] 19.9 or less, adult  |
| Z68.20         | Body mass index \[BMI\] 20.0-20.9, adult     |
| Z68.21         | Body mass index \[BMI\] 21.0-21.9, adult     |
| Z68.22         | Body mass index \[BMI\] 22.0-22.9, adult     |
| Z68.23         | Body mass index \[BMI\] 23.0-23.9, adult     |
| Z68.24         | Body mass index \[BMI\] 24.0-24.9, adult     |
| Z68.25         | Body mass index \[BMI\] 25.0-25.9, adult     |
| Z68.26         | Body mass index \[BMI\] 26.0-26.9, adult     |
| Z68.27         | Body mass index \[BMI\] 27.0-27.9, adult     |
| Z68.28         | Body mass index \[BMI\] 28.0-28.9, adult     |
| Z68.29         | Body mass index \[BMI\] 29.0-29.9, adult     |
| Z68.30         | Body mass index \[BMI\] 30.0-30.9, adult     |
| Z68.31         | Body mass index \[BMI\] 31.0-31.9, adult     |
| Z68.32         | Body mass index \[BMI\] 32.0-32.9, adult     |
| Z68.33         | Body mass index \[BMI\] 33.0-33.9, adult     |
| Z68.34         | Body mass index \[BMI\] 34.0-34.9, adult     |
| Z68.35         | Body mass index \[BMI\] 35.0-35.9, adult     |
| Z68.36         | Body mass index \[BMI\] 36.0-36.9, adult     |
| Z68.37         | Body mass index \[BMI\] 37.0-37.9, adult     |
| Z68.38         | Body mass index \[BMI\] 38.0-38.9, adult     |
| Z68.39         | Body mass index \[BMI\] 39.0-39.9, adult     |
| Z68.41         | Body mass index \[BMI\] 40.0-44.9, adult     |
| Z68.42         | Body mass index \[BMI\] 45.0-49.9, adult     |
| Z68.43         | Body mass index \[BMI\] 50.0-59.9, adult     |
| Z68.44         | Body mass index \[BMI\] 60.0-69.9, adult     |

<br>

``` r
NHANES::NHANES |> 
  dplyr::slice_sample(n = 30) |> 
  janitor::clean_names() |> 
  dplyr::filter(age >= 20) |> 
  dplyr::select(age, weight, height) |> 
  dplyr::mutate(
    weight = units::set_units(weight, kg),
    height = units::set_units(height, cm),
    body_mass_index = bmi_adult(weight = weight, height = height),
    icd_10_code = assign_bmi_icd(body_mass_index),
    bmi_status = assign_bmi_status(body_mass_index)) |> 
  gluedown::md_table()
```

| age |       weight |       height |     body_mass_index | icd_10_code | bmi_status     |
|----:|-------------:|-------------:|--------------------:|:------------|:---------------|
|  62 |  99.0 \[kg\] | 166.1 \[cm\] | 35.88359 \[kg/m^2\] | Z68.35      | Obese          |
|  79 | 103.6 \[kg\] | 172.6 \[cm\] | 34.77588 \[kg/m^2\] | Z68.34      | Obese          |
|  47 |  87.6 \[kg\] | 174.7 \[cm\] | 28.70241 \[kg/m^2\] | Z68.28      | Overweight     |
|  68 |  72.7 \[kg\] | 159.2 \[cm\] | 28.68457 \[kg/m^2\] | Z68.28      | Overweight     |
|  52 |  58.7 \[kg\] | 146.4 \[cm\] | 27.38772 \[kg/m^2\] | Z68.27      | Overweight     |
|  53 |  87.5 \[kg\] | 163.7 \[cm\] | 32.65207 \[kg/m^2\] | Z68.32      | Obese          |
|  61 |  71.5 \[kg\] | 160.6 \[cm\] | 27.72139 \[kg/m^2\] | Z68.27      | Overweight     |
|  73 |  66.8 \[kg\] | 157.3 \[cm\] | 26.99722 \[kg/m^2\] | Z68.26      | Overweight     |
|  49 |  74.6 \[kg\] | 177.1 \[cm\] | 23.78492 \[kg/m^2\] | Z68.23      | Healthy Weight |
|  59 |  95.4 \[kg\] | 173.7 \[cm\] | 31.61904 \[kg/m^2\] | Z68.31      | Obese          |
|  24 |  66.8 \[kg\] | 164.4 \[cm\] | 24.71570 \[kg/m^2\] | Z68.24      | Healthy Weight |
|  57 |  81.1 \[kg\] | 165.9 \[cm\] | 29.46647 \[kg/m^2\] | Z68.29      | Overweight     |
|  52 |  83.9 \[kg\] | 170.4 \[cm\] | 28.89501 \[kg/m^2\] | Z68.28      | Overweight     |
|  31 | 136.3 \[kg\] | 180.9 \[cm\] | 41.65036 \[kg/m^2\] | Z68.41      | Obese          |
|  70 |  84.0 \[kg\] | 180.5 \[cm\] | 25.78249 \[kg/m^2\] | Z68.25      | Overweight     |
|  52 |  39.3 \[kg\] | 144.3 \[cm\] | 18.87382 \[kg/m^2\] | Z68.1       | Healthy Weight |
|  76 |  69.4 \[kg\] | 150.5 \[cm\] | 30.63984 \[kg/m^2\] | Z68.30      | Obese          |
|  32 |  87.2 \[kg\] | 177.1 \[cm\] | 27.80221 \[kg/m^2\] | Z68.27      | Overweight     |
|  52 |  55.8 \[kg\] | 160.7 \[cm\] | 21.60740 \[kg/m^2\] | Z68.21      | Healthy Weight |

<br> <br>

------------------------------------------------------------------------

## NLM’s MedlinePlus Connect API

``` r
rx <- codex_medline(code_title = "Chantix 0.5 MG Oral Tablet", 
                    code_system = "rxcui")

rx |> dplyr::mutate(dplyr::across(everything(), as.character)) |> 
      tidyr::pivot_longer(cols = dplyr::everything()) |> 
      gluedown::md_table()
```

| name                | value                                                                                                                                                                                                                                           |
|:--------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| updated             | 2023-01-16 16:23:23                                                                                                                                                                                                                             |
| entry_title         | Varenicline                                                                                                                                                                                                                                     |
| source              | U.S. National Library of Medicine                                                                                                                                                                                                               |
| search_description  | , RXNORM, Chantix 0.5 MG Oral Tablet, PAT                                                                                                                                                                                                       |
| results_description | MedlinePlus Connect results for RXCUI                                                                                                                                                                                                           |
| entry_link          | <https://medlineplus.gov/druginfo/meds/a606024.html?utm_source=mplusconnect&utm_medium=service>                                                                                                                                                 |
| entry_summary       | Varenicline is used along with education and counseling to help people stop smoking. Varenicline is in a class of medications called smoking cessation aids. It works by blocking the pleasant effects of nicotine (from smoking) on the brain. |

<br> <br>

------------------------------------------------------------------------

## Medicare Fee-for-Service Comprehensive Error Rate Testing API

``` r
codex_cert(hcpcs = "81595", 
           decision = "Disagree") |> 
  dplyr::select(-type_of_bill, 
                -drg, 
                -claim_control_number) |> 
  gluedown::md_table()
```

| year | part       | hcpcs_procedure_code | provider_type                               | review_decision | error_code                 |
|-----:|:-----------|:---------------------|:--------------------------------------------|:----------------|:---------------------------|
| 2021 | 1\. Part B | 81595                | Clinical Laboratory (Billing Independently) | Disagree        | Insufficient Documentation |
| 2021 | 1\. Part B | 81595                | Clinical Laboratory (Billing Independently) | Disagree        | Insufficient Documentation |
| 2021 | 1\. Part B | 81595                | Clinical Laboratory (Billing Independently) | Disagree        | Insufficient Documentation |
| 2021 | 1\. Part B | 81595                | Clinical Laboratory (Billing Independently) | Disagree        | Insufficient Documentation |
| 2021 | 1\. Part B | 81595                | Clinical Laboratory (Billing Independently) | Disagree        | Insufficient Documentation |
| 2021 | 1\. Part B | 81595                | Clinical Laboratory (Billing Independently) | Disagree        | Insufficient Documentation |
| 2021 | 1\. Part B | 81595                | Clinical Laboratory (Billing Independently) | Disagree        | Insufficient Documentation |
| 2021 | 1\. Part B | 81595                | Clinical Laboratory (Billing Independently) | Disagree        | Insufficient Documentation |
| 2021 | 1\. Part B | 81595                | Clinical Laboratory (Billing Independently) | Disagree        | Insufficient Documentation |
| 2021 | 1\. Part B | 81595                | Clinical Laboratory (Billing Independently) | Disagree        | Insufficient Documentation |

<br> <br>

------------------------------------------------------------------------

## NCCI Procedure-to-Procedure Edits (PTP) API

``` r
codex_ptp(column_1 = "29000") |> 
  dplyr::select(-record_number) |> 
  head() |> 
  gluedown::md_table()
```

| quarter_begin_date | category     | column_1 | column_2 | effective_date | deletion_date | modifier_indicator | ptp_edit_rationale            |
|:-------------------|:-------------|:---------|:---------|:---------------|:--------------|:-------------------|:------------------------------|
| 2022-10-01         | DME Services | 29000    | 29010    | 2013-04-01     | NA            | 0                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29000    | 29015    | 2013-04-01     | NA            | 0                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29000    | 29020    | 2013-04-01     | 2014-12-31    | 0                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29000    | 29025    | 2013-04-01     | 2014-12-31    | 0                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29000    | 29035    | 2013-04-01     | NA            | 0                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29000    | 29040    | 2013-04-01     | NA            | 0                  | Mutually exclusive procedures |

<br>

``` r
codex_ptp(column_2 = "L1951") |> 
  dplyr::select(-record_number) |> 
  head() |> 
  gluedown::md_table()
```

| quarter_begin_date | category     | column_1 | column_2 | effective_date | deletion_date | modifier_indicator | ptp_edit_rationale            |
|:-------------------|:-------------|:---------|:---------|:---------------|:--------------|:-------------------|:------------------------------|
| 2022-10-01         | DME Services | 29305    | L1951    | 2013-04-01     | NA            | 1                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29325    | L1951    | 2013-04-01     | NA            | 0                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29345    | L1951    | 2013-04-01     | NA            | 1                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29355    | L1951    | 2013-04-01     | NA            | 1                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29358    | L1951    | 2013-04-01     | NA            | 1                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29365    | L1951    | 2013-04-01     | NA            | 1                  | Mutually exclusive procedures |

<br>

``` r
codex_ptp(category = "DME Services") |> 
  dplyr::select(-record_number) |> 
  head() |> 
  gluedown::md_table()
```

| quarter_begin_date | category     | column_1 | column_2 | effective_date | deletion_date | modifier_indicator | ptp_edit_rationale            |
|:-------------------|:-------------|:---------|:---------|:---------------|:--------------|:-------------------|:------------------------------|
| 2022-10-01         | DME Services | 29000    | 29010    | 2013-04-01     | NA            | 0                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29000    | 29015    | 2013-04-01     | NA            | 0                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29000    | 29020    | 2013-04-01     | 2014-12-31    | 0                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29000    | 29025    | 2013-04-01     | 2014-12-31    | 0                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29000    | 29035    | 2013-04-01     | NA            | 0                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29000    | 29040    | 2013-04-01     | NA            | 0                  | Mutually exclusive procedures |

<br>

``` r
codex_ptp(modifier_indicator = "0") |> 
  dplyr::select(-record_number) |> 
  head() |> 
  gluedown::md_table()
```

| quarter_begin_date | category     | column_1 | column_2 | effective_date | deletion_date | modifier_indicator | ptp_edit_rationale            |
|:-------------------|:-------------|:---------|:---------|:---------------|:--------------|:-------------------|:------------------------------|
| 2022-10-01         | DME Services | 29000    | 29010    | 2013-04-01     | NA            | 0                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29000    | 29015    | 2013-04-01     | NA            | 0                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29000    | 29020    | 2013-04-01     | 2014-12-31    | 0                  | Mutually exclusive procedures |
| 2022-10-01         | DME Services | 29000    | 29025    | 2013-04-01     | 2014-12-31    | 0                  | Mutually exclusive procedures |

NEWS.md
<!-- NEWS.md is maintained by https://cynkra.github.io/fledge, do not edit -->

# codexchain 0.0.0.9004 (2023-01-16)

* codex_ptp() and codex_mue() now use the "SQL API" format


# codexchain 0.0.0.9003 (2023-01-08)

- Same as previous version.


# codexchain 0.0.0.9002 (2023-01-06)

- Same as previous version.


# codexchain 0.0.0.9001 (2023-01-06)

- Same as previous version.




R/utils.R
#' Remove NULL elements from vector ----------------------------------------
#' @autoglobal
#' @noRd
remove_null <- function(l) {Filter(Negate(is.null), l)}

#' param_format ------------------------------------------------------------
#' @param param API parameter
#' @param arg API function arg
#' @return formatted API filters
#' @autoglobal
#' @noRd
param_format <- function(param, arg) {
  if (is.null(arg)) {param <- NULL} else {
    paste0("filter[", param, "]=", arg, "&")}}

#' param_space --------------------------------------------------------------
#' Some API parameters have spaces, these must be converted to "%20".
#' @param param parameter with a space
#' @return parameter formatted with "%20" in lieu of a space
#' @autoglobal
#' @noRd
param_space <- function(param) {gsub(" ", "%20", param)}

#' param_brackets -----------------------------------------------------------
#' Some API parameters have spaces, these must be converted to "%20".
#' @param param parameter with a space
#' @return parameter formatted with "%20" in lieu of a space
#' @autoglobal
#' @noRd
param_brackets <- function(param,
                           right = TRUE,
                           left = TRUE,
                           asterisk = TRUE) {
  param <- gsub("[", "%5B", param, fixed = TRUE)
  param <- gsub("*", "%2A", param, fixed = TRUE)
  param <- gsub("]", "%5D", param, fixed = TRUE)
  return(param)
  }

#' param_format ------------------------------------------------------------
#' @param param API parameter
#' @param arg API function arg
#' @return formatted API filters
#' @autoglobal
#' @noRd
medline_format <- function(param, arg) {
  if (is.null(arg)) {param <- NULL} else {
    x <- gsub(" ", "%20", arg)
    paste0(param, x, "&")}}

#' sql_format ------------------------------------------------------------
#' @param param API parameter
#' @param arg API function arg
#' @return formatted API filters
#' @autoglobal
#' @noRd
sql_format <- function(param, arg) {
  if (is.null(arg)) {param <- NULL} else {
    paste0("[WHERE ", param, " = ", "%22", arg, "%22", "]")}}


#' assign_bmi_icd ----------------------------------------------------------
#' @param bmi body mass index
#' @return vector of assigned icd bmi codes
#' @autoglobal
#' @export
assign_bmi_icd <- function(bmi) {
  bmi <- as.numeric(bmi)
  dplyr::case_when(bmi >= 70 ~ "Z68.45",
                   bmi <= 19.99999 ~ "Z68.1",
                   bmi >= 20.0 & bmi <= 20.99999 ~ "Z68.20",
                   bmi >= 21.0 & bmi <= 21.99999 ~ "Z68.21",
                   bmi >= 22.0 & bmi <= 22.99999 ~ "Z68.22",
                   bmi >= 23.0 & bmi <= 23.99999 ~ "Z68.23",
                   bmi >= 24.0 & bmi <= 24.99999 ~ "Z68.24",
                   bmi >= 25.0 & bmi <= 25.99999 ~ "Z68.25",
                   bmi >= 26.0 & bmi <= 26.99999 ~ "Z68.26",
                   bmi >= 27.0 & bmi <= 27.99999 ~ "Z68.27",
                   bmi >= 28.0 & bmi <= 28.99999 ~ "Z68.28",
                   bmi >= 29.0 & bmi <= 29.99999 ~ "Z68.29",
                   bmi >= 30.0 & bmi <= 30.99999 ~ "Z68.30",
                   bmi >= 31.0 & bmi <= 31.99999 ~ "Z68.31",
                   bmi >= 32.0 & bmi <= 32.99999 ~ "Z68.32",
                   bmi >= 33.0 & bmi <= 33.99999 ~ "Z68.33",
                   bmi >= 34.0 & bmi <= 34.99999 ~ "Z68.34",
                   bmi >= 35.0 & bmi <= 35.99999 ~ "Z68.35",
                   bmi >= 36.0 & bmi <= 36.99999 ~ "Z68.36",
                   bmi >= 37.0 & bmi <= 37.99999 ~ "Z68.37",
                   bmi >= 38.0 & bmi <= 38.99999 ~ "Z68.38",
                   bmi >= 39.0 & bmi <= 39.99999 ~ "Z68.39",
                   bmi >= 40.0 & bmi <= 44.99999 ~ "Z68.41",
                   bmi >= 45.0 & bmi <= 49.99999 ~ "Z68.42",
                   bmi >= 50.0 & bmi <= 59.99999 ~ "Z68.43",
                   bmi >= 60.0 & bmi <= 69.99999 ~ "Z68.44",
                   TRUE ~ "ERROR")
}

#' assign_bmi_status -------------------------------------------------------
#' @param bmi body mass index
#' @return vector of assigned bmi statuses
#' @autoglobal
#' @export
assign_bmi_status <- function(bmi) {
  bmi <- as.numeric(bmi)
  dplyr::case_when(bmi < 18.55 ~ "Underweight",
                   bmi >= 18.55 & bmi <= 24.99999 ~ "Healthy Weight",
                   bmi >= 25.04 & bmi <= 29.99999 ~ "Overweight",
                   bmi >= 30 ~ "Obese",
                   TRUE ~ "ERROR")
}

#' Calculate Body Mass Index (BMI) for Adults
#' @param weight weight in kg
#' @param height height in cm
#' @return vector of BMI values (kg/m2)
#' @examples
#' bmi_adult(70, 160)
#' @autoglobal
#' @export
bmi_adult <- function(weight, height) {

  weight <- units::set_units(weight, kg)
  height <- units::set_units(height, cm)

  units(height) <- units::make_units(m)

  bmi <- weight / (height ^ 2)

  units(bmi) <- units::make_units(kg/m^2)

  return(bmi)

}

#' Calculate Body Mass Index (BMI)
#' @param wt weight in kg
#' @param ht height in cm
#' @return vector of BMI values (kg/m2)
#' @examples
#' calc_bmi(wt = 70, ht = 160)
#' @autoglobal
#' @noRd
calc_bmi <- function(wt, ht) {
  wt / (ht / 100) ^ 2
  }

#' Convert Inches (in) to Centimeters (cm)
#' @param inch inches
#' @examples
#' conv_in_cm(1)
#' @autoglobal
#' @noRd
conv_in_cm <- function(inch) {inch * 2.54}

#' Convert Centimeters (cm) to Inches (in)
#' @param cent centimeters
#' @examples
#' conv_cm_in(2.54)
#' @autoglobal
#' @noRd
conv_cm_in <- function(cent) {cent / 2.54}

#' Convert Kilograms (kg) to Pounds (lbs)
#' @param kg kilograms
#' @examples
#' conv_kg_lbs(1)
#' @autoglobal
#' @noRd
conv_kg_lbs <- function(kg) {kg * 2.20462}

#' Convert Pounds (lbs) to Kilograms (kg)
#' @param lbs pounds
#' @examples
#' conv_lbs_kg(2.20462)
#' @autoglobal
#' @noRd
conv_lbs_kg <- function(lbs) {lbs / 2.20462}

#' Convert Kilograms (kg) to Ounces (oz)
#' @param kg kilograms
#' @examples
#' conv_kg_oz(1)
#' @autoglobal
#' @noRd
conv_kg_oz <- function(kg) {kg * 35.274}

#' Convert Ounces (oz) to Kilograms (kg)
#' @param oz ounces
#' @examples
#' oz2kg(2.20462)
#' @autoglobal
#' @noRd
conv_oz_kg <- function(oz) {oz / 35.274}

#' Percentile BMI for age for children
#' Based on tables from WHO:
#'    http://www.who.int/growthref/who2007_bmi_for_age/en/
#' @param age age in years
#' @param sex either `male` or `female`
#' @param height height
#' @param bmi bmi Optional, if specified, will calculate closest percentile
#'    and return in list as `percentile`
#' @param return_median just return the median expected value
#' @param ... parameters passed to `read_who_table()`
#' @examples
#' pct_bmi_for_age(age = 8, sex = "male")
#' pct_bmi_for_age(age = 8, bmi = 15, sex = "male")
#' @autoglobal
#' @noRd
pct_bmi_for_age <- function(age = NULL,
                            bmi = NULL,
                            sex = NULL,
                            height = NULL,
                            return_median = FALSE,
                            ...) {
  if (is.null(age)) {stop("Age required.")}
  if (length(age) == 1) {pct <- pct_for_age_generic(age = age,
                                                    value = bmi,
                                                    sex = sex,
                                                    variable = "bmi",
                                                    height = height,
                                                    ...)
  } else {
    tmp <- lapply(age, pct_bmi_for_age, sex = sex)
    pct <- data.frame(cbind("age" = age,
                            matrix(unlist(tmp),
                                   nrow = length(age))))
    colnames(pct)[-1] <- names(tmp[[1]])
  }
  if (return_median) {return(pct$P50)
  } else {
    return(pct)
  }
  return(pct)
}

#' Percentile height or weight for age for children
#'
#' This is the underlying function, the exposed functions are
#'    pct_weight_for_age() and pct_height_for_age()
#' Based on tables from WHO:
#'    http://www.who.int/childgrowth/standards/height_for_age/en/
#'
#' @param age age in years
#' @param sex either `male` or `female`
#' @param value height in kg. Optional, if specified, will calculate closest
#'    percentile and return in list as `percentile`
#' @param variable weight or height?
#' @param ... parameters passed to `read_who_table()`
#' @autoglobal
#' @noRd
pct_for_age_generic <- function(age = NULL,
                                value = NULL,
                                sex = NULL,
                                variable = "weight",
                                ...) {
  if (is.null(age) || is.null(sex)) {
    stop("Age and sex are required!")
  }
  if (variable == "height" & age > 19) {
    message("Sorry, height data currently only available for age <= 19 years.")
    return(NULL)
  }
  if (variable == "weight" & age > 10) {
    message("Sorry, currently only available for age <= 10 years.")
    return(NULL)
  }

  if (variable == "height") {
    type = ifelse(age >= 5.1, "hfa", "lhfa")
  } else if (variable == "bmi") {
    type = ifelse(age >= 5.1, "bmi", "bfa")
  } else {
    type <- "wfa"
  }

  dat <- read_who_table(sex = sex, age = age, type = type)
  tmp <- dat[which.min(abs(age - dat$age)),-(1:4)]
  pct <- as.list(tmp)
  if (!is.null(value)) {
    p <- c()
    for (i in seq(names(pct))) {
      p <- c(p, as.num(gsub("P", "", names(pct)[i])))
    }
    p[1] <- p[1]/10 # 0.1
    p[length(p)] <- p[length(p)]/10 # 99.9
    p_txt <- paste0("pct_", p)
    if (value >= max(tmp)) {
      message(paste0("Specified ", variable," >= 99.9th percentile!"))
      pct <- list(percentile = 99.9)
    }
    if (value <= min(tmp)) {
      message(paste0("Specified ", variable, " <= 0.1th percentile!"))
      pct <- list(percentile = 0.1)
    }
    if (is.null(pct$percentile)) {
      data <- data.frame(
        x = c(
          as.num(tmp[value <= as.num(tmp)][1]),
          tail(as.num(tmp[value > as.num(tmp)]),1)
        ),
        y = c(
          p[value <= as.num(tmp)][1],
          tail(p[value > as.num(tmp)],1)
        )
      )
      # linearly scale between two values
      fit <- lm(y~x, data)
      par <- coef(fit)
      pct <- list(percentile = round(as.num(par[1] + par[2]*value),1))
    }
  }
  return(pct)
}

#' Percentile height for age for children
#'
#' Based on tables from WHO: http://www.who.int/childgrowth/standards/height_for_age/en/
#'
#' @param age age in years
#' @param sex either `male` or `female`
#' @param height height in kg. Optional, if specified, will calculate closest
#'    percentile and return in list as `percentile`
#' @param return_median just return the median expected value
#' @param ... parameters passed to `read_who_table()`
#' @examples
#' pct_height_for_age(age = 5, sex = "female")
#' pct_height_for_age(age = 5, height = 112, sex = "female")
#' @autoglobal
#' @noRd
pct_height_for_age <- function(age = NULL,
                               height = NULL,
                               sex = NULL,
                               return_median = FALSE,
                               ...) {
  if (is.null(age)) {
    stop("Age required.")
  }
  if (length(age) == 1) {
    if (return_median) {
      pct <- pct_for_age_generic(age = age,
                                 sex = sex,
                                 variable = "height",
                                 ...)
    } else {
      pct <- pct_for_age_generic(age = age,
                                 value = height,
                                 sex = sex,
                                 variable = "height",
                                 ...)
    }
  } else {
    if (is.null(height)) {
      tmp <- lapply(age, pct_height_for_age, sex = sex)
    } else {
      stop("A specific height value cannot be supplied when age is a vector.")
    }
    pct <- data.frame(cbind("age" = age,
                            matrix(unlist(tmp),
                                   nrow = length(age))))
    colnames(pct)[-1] <- names(tmp[[1]])
  }
  if (return_median) {
    return(pct$P50)
  } else {
    return(pct)
  }
}

#' Percentile weight for age for children
#'
#' Based on tables from WHO:
#'    http://www.who.int/childgrowth/standards/weight_for_age/en/
#'
#' @param age age in years
#' @param sex either `male` or `female`
#' @param weight weight in kg. Optional, if specified, will calculate closest
#'    percentile and return in list as `percentile`
#' @param return_median just return the median expected value
#' @param ... parameters passed to `read_who_table()`
#' @examples
#' pct_weight_for_age(age = 5, sex = "female")
#' pct_weight_for_age(age = 5, weight = 20, sex = "female")
#' @autoglobal
#' @noRd
pct_weight_for_age <- function(age = NULL,
                               weight = NULL,
                               sex = NULL,
                               return_median = FALSE,
                               ...) {
  if (is.null(age)) {
    stop("Age required.")
  }
  if (length(age) == 1) {
    if (return_median) {
      pct <- pct_for_age_generic(age = age,
                                 sex = sex,
                                 variable = "weight",
                                 ...)
    } else {
      pct <- pct_for_age_generic(age = age,
                                 value = weight,
                                 sex = sex,
                                 variable = "weight",
                                 ...)
    }
  } else {
    if (is.null(weight)) {
      tmp <- lapply(age, pct_weight_for_age, sex = sex)
    } else {
      stop("A specific weight value cannot be supplied when age is a vector.")
    }
    pct <- data.frame(cbind("age" = age,
                            matrix(unlist(tmp),
                                   nrow = length(age))))
    colnames(pct)[-1] <- names(tmp[[1]])
  }
  if (return_median) {
    return(pct$P50)
  } else {
    return(pct)
  }
}
